relayer/catchup_listener.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package relayer
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	MaxBlocksPerRequest = 200
+	// Max buffer size for subscription channels
+	maxClientSubscriptionBuffer = 20000
+	// Max number of retries for RPC calls before failing catchup
+	maxRpcRetries = 5
+)
+
+func (lstnr *Listener) processFromHeight(height *big.Int, done chan bool) {
+	defer close(done)
+	lstnr.logger.Info(
+		"Processing historical logs",
+		zap.String("fromBlockHeight", height.String()),
+		zap.String("blockchainID", lstnr.sourceBlockchain.BlockchainID),
+	)
+	if height == nil {
+		lstnr.logger.Error("Cannot process logs from nil height")
+		done <- false
+		return
+	}
+
+	// Grab the latest block before filtering logs so we don't miss any before updating the db
+	latestBlockHeight, err := lstnr.ethClient.BlockNumber(context.Background())
+	if err != nil {
+		lstnr.logger.Error(
+			"Failed to get latest block",
+			zap.String("blockchainID", lstnr.sourceBlockchain.BlockchainID),
+			zap.Error(err),
+		)
+		done <- false
+		return
+	}
+
+	bigLatestBlockHeight := big.NewInt(0).SetUint64(latestBlockHeight)
+
+	//nolint:lll
+	for fromBlock := big.NewInt(0).Set(height); fromBlock.Cmp(bigLatestBlockHeight) <= 0; fromBlock.Add(fromBlock, big.NewInt(MaxBlocksPerRequest)) {
+		toBlock := big.NewInt(0).Add(fromBlock, big.NewInt(MaxBlocksPerRequest-1))
+
+		// clamp to latest known block because we've already subscribed
+		// to new blocks and we don't want to double-process any blocks
+		// created after that subscription but before the determination
+		// of this "latest"
+		if toBlock.Cmp(bigLatestBlockHeight) > 0 {
+			toBlock.Set(bigLatestBlockHeight)
+		}
+
+		for i := 0; i < maxRpcRetries; i++ {
```

Comment:
Looking over it I just realized that it would be cleaner if the retries were on the processBlockRange level and closer to the actual network call since failures here would re-trigger reprocessing of block headers that we've already potentially seen

Reply:
Yeah, definitely agree here :)